AnkiDroid/src/main/java/com/ichi2/libanki/Sched.java

```diff
@@ -1205,10 +1208,35 @@ public int _revForDeck(long did, int lim) {
     	return mCol.getDb().queryScalar("SELECT count() FROM (SELECT 1 FROM cards WHERE did = " + did + " AND queue = 2 AND due <= " + mToday + " LIMIT " + lim + ")");
     }
 
-    public int _youngRevCountForDeck(long did) {
-        return mCol.getDb().queryScalar(
-            "SELECT COUNT(*) FROM cards WHERE did = " + did + " AND queue = 2 AND due <= " +
-                mToday + " AND ivl <= 21");
+    public int _criticalReviewCountForDecks(LinkedList<Long> dids) {
+        int total = 0;
+        for (long did : dids) {
+            total += _criticalReviewCountForDeck(did);
+        }
+        return total;
+    }
+
+    public int _criticalReviewCountForDeck(long did) {
+        return mCol.getDb().queryScalar(getCriticalReviewQueryGivenSelect("COUNT(*)", did));
+    }
+
+    public String getCriticalReviewQueryGivenSelect(String selectString, long did) {
+        // using DISTINCT because our join here is approximate (multiple revlog entries could match
+        // for the same card; using this join approach for performance)
+        return String.format("SELECT DISTINCT %s " +
+                        "FROM cards " +
+                        "JOIN revlog ON revlog.cid = cards.id AND revlog.factor = cards.factor AND revlog.ivl = cards.ivl " +
```

Comment:
it is a bummer that we have to perform this join here because it causes a slight performance degradation. It would be good to look into removing some functionality we don't use anyway that also gets called on each new review to offset.

Reply:
is the performance degradation noticeable during review? if so, I'll look into optimizing this.

perf is something we've mentioned addressing a few times in the past. if you think it's high lev, i can run an analysis and try to pick the low-hanging fruit.

Reply:
I did notice it right when I started using the app but now it seems less noticeable, possibly because I'm getting further through the backlog. I think for now we can just leave it as it is, seems fine to me. Though eventually it would be good to look at perf (upstream they are mainly focused on fixing crashes, from what I read)